doc/dynamo0p3.rst

```diff
@@ -1171,27 +1207,71 @@ rules, along with PSyclone's naming conventions, are:
+
+	   1) if ``gh_shape`` is ``gh_quadrature_xyoz`` then the
+	      arrays are of rank 4 with extent (``dimension``,
+	      ``number_of_dofs``, ``np_xy``, ``np_z``). The name of
+	      the argument is ``"basis_"<field_function_space>`` or
+	      ``"diff_basis_"<field_function_space>``, as appropriate.
+
+	   2) if ``gh_shape`` is ``gh_evaluator`` then we pass one
+	      array for each target function space (e.g\. as specified
+	      by ``gh_evaluator_targets``).  Each of these arrays are
+	      of rank 3 with extent (``dimension``,
+	      ``number_of_dofs``,
+	      ``ndf_<target_function_space>``). The name of the
+	      argument is
+	      ``"basis_"<field_function_space>"_on_"<target_function_space>``
+	      or
+	      ``"diff_basis_"<field_function_space>"_on_"<target_function_space>``,
+	      as appropriate.
 	      
-           where ``dimension`` is 1 or 3 and depends upon the function space and whether or not it is a basis or a differential basis function. For the former it is (w0=1, w1=3, w2=3, w3=1, wtheta=1, w2h=3, w2v=3, any_w2=3). For the latter it is (w0=3, w1=3, w2=1, w3=3, wtheta=3, w2h=1, w2v=1, any_w2=3). ``number_of_dofs`` is the number of degrees of freedom (dofs) associated with the function space and ``np_*`` are the number of points to be evaluated: i) ``*_xyz`` in all directions (3D); ii) ``*_xy`` in the horizontal plane (2D); iii) ``*_x, *_y`` in the horizontal (1D); and iv) ``*_z`` in the vertical (1D).
-
-        2) If it is an orientation array, include the associated argument. The argument is an integer array with intent ``in``. There is one dimension of size the local degrees of freedom for the function space. The name of the array is ``"orientation_"<field_function_space>``.
-
-5) If Quadrature is required (``gh_shape = gh_quadrature_*``)
+           where ``dimension`` is 1 or 3 and depends upon the function
+           space and whether or not it is a basis or a differential
+           basis function. For the former it is (w0=1, w1=3, w2=3,
+           w3=1, wtheta=1, w2h=3, w2v=3, any_w2=3). For the latter it
+           is (w0=3, w1=3, w2=1, w3=3, wtheta=3, w2h=1, w2v=1,
+           any_w2=3). ``number_of_dofs`` is the number of degrees of
+           freedom (dofs) associated with the function space and
+           ``np_*`` are the number of points to be evaluated: i)
+           ``*_xyz`` in all directions (3D); ii) ``*_xy`` in the
+           horizontal plane (2D); iii) ``*_x, *_y`` in the horizontal
+           (1D); and iv) ``*_z`` in the vertical (1D).
+
+        2) If it is an orientation array, include the associated
+           argument. The argument is an integer array with intent
+           ``in``. There is one dimension of size the local degrees of
+           freedom for the function space. The name of the array is
+           ``"orientation_"<field_function_space>``.
+	   
+5) If either the ``normal_to_face`` or ``out_face_normal`` properties
```

Comment:
As explained above, number of faces may not only be horizontal. Also, it can come from the face quadrature object. I would suggest something like
```
6) If ``adjacent_face`` property of the mesh is required then
    pass 
   number of faces 
 must be passed as well. There are two LFRic objects to obtain the number
 of faces from: face quadrature and reference element. In latter case,
 in the order specified in the ``meta_reference_element`` metadata:

   1) For ``number_horizontal_faces`` or ``number_vertical_faces`` pass
      the ``nfaces_re_h`` or ``nfaces_re_v``  integer, respectively.
   2) For ``normal_to_face``, pass a rank-2 integer array
      with dimensions ``(nfaces_re_*, 3)``, where ``*`` can be ``h`` or ``v``
      as specified above.
   2) For ``out_face_normal``, pass a rank-2 integer array with dimensions
      ``(3, nfaces_re_*)`` where ``*`` can be ``h`` or ``v`` as specified above.
```
We will also need rules for the `meta_mesh` properties, so I would suggest something like 
```
A kernel that requires properties of the LFRic mesh object specifies
those properties through the ``meta_mesh`` metadata entry. (If no
mesh properties are required then this metadata should be omitted.)
Consider the following example kernel metadata::

  type, extends(kernel_type) :: testkern_type
    type(arg_type), dimension(2) :: meta_args = &
        (/ arg_type(gh_field, gh_read, w1),     &
	   arg_type(gh_field, gh_write, w0) /)
    type(mesh_data_type), dimension(1) ::       &
      meta_mesh =                               &
        (/ mesh_data_type(adjacent_face) /)
  contains
    procedure, nopass :: code => testkern_code
  end type testkern_type

This metadata specifies that the ``testkern_type`` kernel requires one
property of the mesh. There is currently one supported property:
```